token-metadata/program/src/processor/verification/collection_verification.rs

```diff
@@ -0,0 +1,218 @@
+    let mut metadata = Metadata::from_account_info(ctx.accounts.metadata_info)?;
+
+    // Destructure the collection field from the item metadata.  If there's no collection set, we
+    // can just short-circuit since there's nothing to unverify.
+    let collection = match metadata.collection.as_mut() {
+        Some(collection) => collection,
+        None => return Ok(()),
+    };
+
+    // Short-circuit if it's already unverified.
+    if !collection.verified {
+        return Ok(());
+    }
+
+    // The collection parent must be the actual parent of the collection item.
+    if collection.key != *collection_mint_info.key {
+        return Err(MetadataError::NotAMemberOfCollection.into());
+    }
+
+    // Ensure the metadata is derived from the mint.
+    assert_metadata_derivation(program_id, collection_metadata_info, collection_mint_info)?;
+
+    // Check if the collection metadata account is burned. If it is, there's no sized data to
+    // update and the user can simply unverify the NFT.
+    let parent_burned = collection_metadata_info.data_is_empty();
+
+    if parent_burned {
+        // If the collection parent is burned, we need to use an authority for the item rather than
+        // the collection.  The required authority is either the item's metadata update authority,
+        // or an update delegate for the item.  This call fails if no valid authority is present.
+        let authority_response = AuthorityType::get_authority_type(AuthorityRequest {
+            authority: ctx.accounts.authority_info.key,
+            update_authority: &metadata.update_authority,
+            mint: &metadata.mint,
+            metadata_delegate_record_info: ctx.accounts.delegate_record_info,
+            metadata_delegate_roles: vec![MetadataDelegateRole::Update],
+            precedence: &[AuthorityType::Metadata, AuthorityType::MetadataDelegate],
+            ..Default::default()
+        })?;
+
+        // Validate that authority type is expected.
+        match authority_response.authority_type {
+            AuthorityType::Metadata | AuthorityType::MetadataDelegate => (),
+            _ => return Err(MetadataError::UpdateAuthorityIncorrect.into()),
+        }
+    } else {
+        // If the parent is not burned, we need to ensure the collection metadata account is owned
+        // by the token metadata program.
+        assert_owned_by(collection_metadata_info, program_id)?;
+
+        // Now we can deserialize the collection metadata account.
+        let mut collection_metadata = Metadata::from_account_info(collection_metadata_info)?;
+
+        // Determines if we have a valid authority to perform the collection verification.  The
+        // required authority is either the collection parent's metadata update authority, or a
+        // collection delegate for the collection parent.  This call fails if no valid authority is
+        // present.
+        let authority_response = AuthorityType::get_authority_type(AuthorityRequest {
```

Comment:
[I think there's a refactor here that reduces code redundancy](https://github.com/metaplex-foundation/metaplex-program-library/pull/1043), but you may not prefer it or see something better so feel free to not use the PR.

Reply:
Merged it but currently getting an issue with: error[E0597]: `collection_metadata.update_authority` does not live long enough`.  Also I need to specify different `metadata_delegate_roles` for the two cases.  Let me take a look at it.

Reply:
Ok fixed the issues and incorporated this.